[Cocoa] Prevent active background page processes from suspending

salinas-miguel opened the pull request:
#### 06fe365b16021e6ae0c4a3b66e5e8351de765902
<pre>
[Cocoa] Prevent active background page processes from suspending
<a href="https://bugs.webkit.org/show_bug.cgi?id=252415">https://bugs.webkit.org/show_bug.cgi?id=252415</a>
rdar://105413250

Reviewed by Chris Dumez.

When a page is backgrounded we try to suspend its process if possible. When the
internal setting &quot;Take WebKit:Suspended assertion...&quot; is disabled, this will
cause the UI process to drop all of its process assertions on the web content
process, effectively making it un-schedulable. This means that pages which may
still be active will completely stop. This wasn&apos;t an issue before since the web
content process would still be schedulable because it was holding the
WebKit:Suspended assertion (this is very confusing since the WebKit:Suspended
process assertion actually _prevents_ suspension). Instead of dropping all
process assertions immediately upon being backgrounded, we should give the
page/process some time to be scheduled and if it performs any observable work
(e.g. changing document title text, sending a local notification, etc.) we
should delay suspending the process.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::assertionTypeForState):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateIfNeeded):
(WebKit::ProcessThrottler::removeAllAssertionsTimerFired):
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):
(WebKit::ProcessThrottler::delaySuspension):
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didReceiveTitleForFrame):
(WebKit::WebPageProxy::showNotification):

Canonical link: <a href="https://commits.webkit.org/260482@main">https://commits.webkit.org/260482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af493b3fe63d689da10d7f8a9e5a227051c44f20

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117523 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116886 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18959 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8783 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100635 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114179 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97440 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29085 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10332 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30429 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11078 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50025 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7252 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12664 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

salinas-miguel commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -39,6 +39,7 @@
 namespace WebKit {
     
 static const Seconds processSuspensionTimeout { 20_s };
+static const Seconds removeAllAssertionsTimeout { 500_s };
```
This is an arbitrary number. I'm not sure what a good delay would be.

cdumez replied:
I like `8_min` which is magic value we use a lot at Apple for this sort of things.

salinas-miguel commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -320,10 +326,15 @@ void ProcessThrottler::setAllowsActivities(bool allow)
 void ProcessThrottler::setShouldTakeSuspendedAssertion(bool shouldTakeSuspendedAssertion)
 {
-    const bool shouldUpdateAssertion = m_shouldTakeSuspendedAssertion != shouldTakeSuspendedAssertion;
     m_shouldTakeSuspendedAssertion = shouldTakeSuspendedAssertion;
```
Actually, I think I should rename this. Maybe `m_canDropSuspendedAssertion`

salinas-miguel replied:
I'll rename in a follow up

salinas-miguel commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5527,6 +5527,8 @@ void WebPageProxy::didReceiveTitleForFrame(FrameIdentifier frameID, const String
     
     m_pageLoadState.commitChanges();
 
+    process().throttler().delaySuspension();
```
Not sure why I didn't just use `m_process`. Changing that now.

cdumez replied:
process() looks safer since it returns a reference. It makes it clear that the process can't be null. `m_process->` looks like m_process might be null. It is minor though.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -320,10 +326,15 @@ void ProcessThrottler::setAllowsActivities(bool allow)
+    PROCESSTHROTTLER_RELEASE_LOG("delaySuspension");
+    if (m_removeAllAssertionsTimer.isActive())
+        m_removeAllAssertionsTimer.startOneShot(removeAllAssertionsTimeout);
+    return;
```
not needed.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9308,6 +9310,7 @@ void WebPageProxy::requestNotificationPermission(const String& originString, Com
 void WebPageProxy::showNotification(IPC::Connection& connection, const WebCore::NotificationData& notificationData, RefPtr<WebCore::NotificationResources>&& notificationResources)
 {
     m_process->processPool().supplement<WebNotificationManagerProxy>()->show(this, connection, notificationData, WTFMove(notificationResources));
+    m_process->throttler().delaySuspension();
```
I don't understand this? how would a background page even request the notification permission? AFAIK, this requires a user gesture.

salinas-miguel replied:
This is `WebPageProxy::showNotification` not `WebPageProxy::requestNotificationPermission`

cdumez replied:
Oh, got confused by the diff. Never mind.

salinas-miguel committed:
[Cocoa] Prevent active background page processes from suspending
https://bugs.webkit.org/show_bug.cgi?id=252415
rdar://105413250

Reviewed by Chris Dumez.

When a page is backgrounded we try to suspend its process if possible. When the
internal setting "Take WebKit:Suspended assertion..." is disabled, this will
cause the UI process to drop all of its process assertions on the web content
process, effectively making it un-schedulable. This means that pages which may
still be active will completely stop. This wasn't an issue before since the web
content process would still be schedulable because it was holding the
WebKit:Suspended assertion (this is very confusing since the WebKit:Suspended
process assertion actually _prevents_ suspension). Instead of dropping all
process assertions immediately upon being backgrounded, we should give the
page/process some time to be scheduled and if it performs any observable work
(e.g. changing document title text, sending a local notification, etc.) we
should delay suspending the process.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::assertionTypeForState):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateIfNeeded):
(WebKit::ProcessThrottler::removeAllAssertionsTimerFired):
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):
(WebKit::ProcessThrottler::delaySuspension):
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didReceiveTitleForFrame):
(WebKit::WebPageProxy::showNotification):

Canonical link: https://commits.webkit.org/260482@main
